### PR TITLE
[LETS-109] Do not read log page from local files on TS with remote storage

### PR DESCRIPTION
**Pull request by cristiarg**
http://jira.cubrid.org/browse/LETS-109

On a transaction server:
-  when `remote_storage` parameter is set, do not read log pages from the local storage anymore
-  when 'remote_storage` parameter is not set, read log pages from both local and page server and, in debug mode, perform a comparison to ensure they are the same (this is temporary code that will be removed later on)
On a page server and in standalone mode read log pages from local storage - no change.

Implementations:
- redirect all calls to `logpb_read_page_from_file` via a new function `logpb_read_page_from_file_or_page_server` that implements the above logic
- the comparison between the log pages read from local disk and those received from page server is expected to be verbatim, except for `LOGPB_HEADER_PAGE_ID` and when the page is actually the current page that log is being appended to
- in `boot_restart_server`, move the initialization of the transaction server's connection with the page server before the log subsystem is initialized; because the latter needs to request log pages from the page server

### CODE REVIEW

File: src/transaction/log_page_buffer.c [hunk 1]
```diff
@@ -1909,13 +1910,15 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
   else
     {
       er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, pageid);
+      // TODO: LOG_CS_EXIT?
```
Comment (cristiarg) [inline src/transaction/log_page_buffer.c]: I think `LOG_CS_EXIT` is needed here?

File: src/transaction/log_page_buffer.c [hunk 2]
```diff
@@ -1999,11 +1981,77 @@ request_log_page_from_ps (LOG_PAGEID log_pageid)
...
+    {
+      std::shared_ptr < log_page_owner > log_page_from_page_server
+	= ats_Gl.get_log_page_broker ().wait_for_page (pageid);
+      if (read_from_disk)
+	{
+	  // context 2)
+	  // argument log_pgptr already contains value read from local storage
+	  const LOG_PAGE *const log_page_from_page_server_pgptr = log_page_from_page_server->get_log_page ();
+	  assert (*log_page_from_page_server_pgptr == *log_pgptr
+		  || pageid == LOGPB_HEADER_PAGE_ID || pageid == log_Gl.hdr.append_lsa.pageid);
```
Comment (cristiarg) [inline src/transaction/log_page_buffer.c]: For `LOGPB_HEADER_PAGE_ID` and `log_Gl.hdr.append_lsa.pageid` we should still add a stronger check - eg: for `hdr` - than simply skipping.

File: src/transaction/log_page_buffer.c [hunk 3]
```diff
@@ -1909,13 +1910,15 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
...
       er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_LOG_PAGE_CORRUPTED, 1, pageid);
+      // TODO: LOG_CS_EXIT?
       return ER_LOG_PAGE_CORRUPTED;
     }
 
   if (log_bufptr->pageid == pageid)
     {
       /* Copy page from log_bufptr into log_pgptr */
       memcpy (log_pgptr, log_bufptr->logpage, LOG_PAGESIZE);
+      // TODO: why this extra check?
```
Comment (cristiarg) [inline src/transaction/log_page_buffer.c]: why?

File: src/transaction/boot_sr.c [hunk 4]
```diff
@@ -2523,6 +2523,14 @@ boot_restart_server (THREAD_ENTRY * thread_p, bool print_restart, const char *db
    * Now restart the recovery manager and execute any recovery actions
    */
 
+#if defined (SERVER_MODE)
+  // transaction server _needs_ connection with page server
+  if (get_server_type () == SERVER_TYPE_TRANSACTION)
+    {
+      ats_Gl.init_log_page_broker ();
+    }
+#endif // SERVER_MODE
```
Comment (ribram) [inline src/transaction/boot_sr.c]: Actually, both page brokers will be initialized before the first boot_mount as of #2823. After that PR is merged, the conflict will have to be resolved.

File: src/transaction/boot_sr.c [hunk 5]
```diff
@@ -2557,15 +2565,13 @@ boot_restart_server (THREAD_ENTRY * thread_p, bool print_restart, const char *db
     }
 
 #if defined (SERVER_MODE)
+  // TODO: not sure, but I think page server's initialization must come after log has been initialized
```
Comment (ribram) [inline src/transaction/boot_sr.c]: The supposition is right. Update the comment :).

File: src/transaction/log_page_buffer.c (hunk 3 above)
Comment (ribram) [reply src/transaction/log_page_buffer.c]: The read/writes of log page buffer entries is not synchronized with locks. If the page is very "old", the cache entry of the page may be overwritten by a new page. If the write happens concurrently with the memcpy, log_bufptr->pageid changes and the read operation from the log page buffer is invalidated.

File: src/transaction/log_page_buffer.c (hunk 1 above)
Comment (ribram) [reply src/transaction/log_page_buffer.c]: Yes.

File: src/transaction/log_page_buffer.c [hunk 6]
```diff
@@ -1999,11 +1981,77 @@ request_log_page_from_ps (LOG_PAGEID log_pageid)
...
+ */
+int
+logpb_read_page_from_file_or_page_server (THREAD_ENTRY * thread_p, LOG_PAGEID pageid,
+					  LOG_CS_ACCESS_MODE access_mode, LOG_PAGE * log_pgptr)
+{
+#if defined (SERVER_MODE)
+  // execution contexts:
+  //  1) TS with remote storage => only read from PS
+  //  2) TS with local storage => both read from PS and from file and compare results
+  //  3) PS  => only read from file
```
Comment (ribram) [inline src/transaction/log_page_buffer.c]:
```suggestion
  //  2) TS with local storage + connected to PS => both read from PS and from file and compare results
  //  3) PS or TS with local storage + not connected to PS  => only read from file
```

File: src/transaction/log_page_buffer.c (hunk 2 above)
Comment (ribram) [reply src/transaction/log_page_buffer.c]: For LOGPB_HEADER_PAGE_ID, hdr may be different.
For log_Gl.hdr.append_lsa.pageid, pages may be checked up to log_Gl.hdr.append_lsa.offset.

If the check for append page is added, it is probably best to make a function for all the checks.

File: src/transaction/log_page_buffer.c [hunk 7]
```diff
@@ -1999,11 +1981,77 @@ request_log_page_from_ps (LOG_PAGEID log_pageid)
...
+	{
+	  ASSERT_ERROR ();
+	  return res_read_from_file;
+	}
+      read_from_disk = true;
+    }
+
+  if (server_type == SERVER_TYPE_TRANSACTION && ats_Gl.is_page_server_connected ())
+    {
+      std::shared_ptr < log_page_owner > log_page_from_page_server
```
Comment (ribram) [inline src/transaction/log_page_buffer.c]: Disable indent.

File: src/transaction/log_impl.h [hunk 8]
```diff
@@ -800,6 +800,8 @@ extern int logpb_fetch_page (THREAD_ENTRY * thread_p, const LOG_LSA * req_lsa, L
 			     LOG_PAGE * log_pgptr);
 extern int logpb_copy_page_from_log_buffer (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_PAGE * log_pgptr);
 extern int logpb_copy_page_from_file (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_PAGE * log_pgptr);
```
Comment (ribram) [inline src/transaction/log_impl.h:802]: Shouldn't logpb_copy_page_from_file be moved to log_page_buffer.c now that it is replaced by logpb_read_page_from_file_or_page_server?

File: src/transaction/log_page_buffer.c (hunk 2 above)
Comment (cristiarg) [reply src/transaction/log_page_buffer.c]: Done

File: src/transaction/boot_sr.c (hunk 4 above)
Comment (cristiarg) [reply src/transaction/boot_sr.c]: Done. Removed my change and left the one in #2823 .

File: src/transaction/boot_sr.c (hunk 5 above)
Comment (cristiarg) [reply src/transaction/boot_sr.c]: Done

File: src/transaction/log_page_buffer.c (hunk 6 above)
Comment (cristiarg) [reply src/transaction/log_page_buffer.c]: Done

File: src/transaction/log_impl.h (hunk 8 above)
Comment (cristiarg) [reply src/transaction/log_impl.h:802]: Done

File: src/transaction/log_page_buffer.c [hunk 9]
```diff
@@ -310,6 +310,13 @@ static void logpb_set_unavailable_archive (THREAD_ENTRY * thread_p, int arv_num)
 static void logpb_dismount_log_archive (THREAD_ENTRY * thread_p);
 static bool logpb_is_archive_available (THREAD_ENTRY * thread_p, int arv_num);
 static void logpb_archive_active_log (THREAD_ENTRY * thread_p);
+#if defined (SERVER_MODE)
+static void logpb_assert_correctness_read_page_from_file_or_page_server (LOG_PAGEID pageid,
```
Comment (ribram) [inline src/transaction/log_page_buffer.c]: This name really screams to be shortened :)). What about just `logpb_verify_page_read`?

File: src/transaction/log_page_buffer.c [hunk 10]
```diff
@@ -1999,11 +1990,97 @@ request_log_page_from_ps (LOG_PAGEID log_pageid)
...
         }
-    }  
+    }
   // *INDENT-ON*
 #endif // SERVER_MODE
 }
 
+#if defined (SERVER_MODE)
+void logpb_assert_correctness_read_page_from_file_or_page_server (LOG_PAGEID pageid,
+    const LOG_PAGE * left_log_pgptr, const LOG_PAGE * rite_log_pgptr)
```
Comment (ribram) [inline src/transaction/log_page_buffer.c]: Coding style for function definition: return type on the first line, the function name and first arguments in the second line.

File: src/transaction/log_page_buffer.c [hunk 11]
```diff
@@ -1999,11 +1990,97 @@ request_log_page_from_ps (LOG_PAGEID log_pageid)
...
+	  // context 2)
+	  // log_pgptr already contains value read from local storage
+	  const LOG_PAGE *const log_page_from_page_server_pgptr = log_page_from_page_server->get_log_page ();
+	  logpb_assert_correctness_read_page_from_file_or_page_server(
+		pageid, log_page_from_page_server_pgptr, log_pgptr);
+	}
+      else
+	{
+	  // context 1)
+	  std::memcpy (log_pgptr, log_page_from_page_server->get_log_page (), LOG_PAGESIZE);
```
Comment (ribram) [inline src/transaction/log_page_buffer.c:2073]: Disable indent for `std::memcpy`

**Review by ribram — APPROVED** [review]
Please merge after making the changes.

File: src/transaction/log_page_buffer.c (hunk 11 above)
Comment (cristiarg) [reply src/transaction/log_page_buffer.c:2073]: Done

File: src/transaction/log_page_buffer.c (hunk 10 above)
Comment (cristiarg) [reply src/transaction/log_page_buffer.c]: Done

File: src/transaction/log_page_buffer.c (hunk 9 above)
Comment (cristiarg) [reply src/transaction/log_page_buffer.c]: Done

**Pull request merged**